# Week 1 & 2 Coding Assignment:

Author: Rosie Navarro

Date: 10 September 2026

Description: 


In [31]:
import random
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from itertools import permutations, dropwhile
import math

There are 5 main steps that need to be completed for step 2: Create a list, iterate over the list, sort results, generate random numbers, add to the list, and print the results. 

The following creates a list of numbers that I chose. 

In [2]:
numbers = [10, 1, 15, 5, 7, 23, 2, 6]
print("Number list is:", numbers)

Number list is: [10, 1, 15, 5, 7, 23, 2, 6]


The main idea of iteration is that the step are done over and over again until the list is complete. In this case, there are a total of 8 numbers, the code will run until the numbers go through the loop.

In [3]:
print("Numbers in the list:")

for number in numbers:
    print(number)

Numbers in the list:
10
1
15
5
7
23
2
6


Numbers can then be sorted, for this example the numbers can be sorted from smallest to biggest.

In [4]:
numbers.sort() # reverse=true if descending order

print(numbers)

[1, 2, 5, 6, 7, 10, 15, 23]


Loop can also be used to create additional numbers for the list. 

In [5]:
for i in range(5):
    numbers.append(random.randint(1,1000))

In [6]:
numbers.sort()

print("Final Number list:", numbers)

Final Number list: [1, 2, 5, 6, 7, 10, 15, 23, 254, 443, 454, 948, 963]


## Workbook Assignments:

a. activity 1.01: the main idea of this activity is to generate a random list of numbers, new list will be made from the previpous list generating another list that has numbers that are divisible by three. These steps will be done 10 times. 

The first step is to generate a list of 50 random values. These values will be between 0 and 100. 

In [7]:
random_list = [random.randint(0, 100) for x in range (0,50)]
# randint is saying the generated values must be intergers, values can be between 0-100
# for statement means that it will be done 50

random_list

[1,
 40,
 75,
 23,
 50,
 40,
 18,
 97,
 63,
 2,
 97,
 24,
 60,
 71,
 63,
 85,
 57,
 35,
 21,
 72,
 43,
 48,
 16,
 3,
 50,
 17,
 48,
 0,
 71,
 23,
 6,
 1,
 73,
 2,
 72,
 33,
 27,
 70,
 61,
 31,
 40,
 85,
 57,
 20,
 78,
 37,
 47,
 63,
 0,
 47]

Now that 50 values were generated, these will be used to make a new list, with values that are divisible by three. 

In [8]:
list_by_three = [a for a in random_list if a % 3 == 0]

list_by_three

[75,
 18,
 63,
 24,
 60,
 63,
 57,
 21,
 72,
 48,
 3,
 48,
 0,
 6,
 72,
 33,
 27,
 57,
 78,
 63,
 0]

Out of the 50 original values, there are only 21 values that are divisible by three. 

The next step is to find the difference in length between the two list. 

In [11]:
length_random = len(random_list)

length_three = len(list_by_three)

print(length_random)
print(length_three)

50
21


In [12]:
difference = length_random - length_three

difference

29

The first list has a total of 50 values, the second has 21. The difference can be found by subtracting these values from each other. There is a difference of 29 values between the two list. 

The previous steps will now be done 10 times. The output should be a total of 10 different values, which show the difference between the two list. 

In [13]:
number_of_times = 10
difference_list = []

# loop
for i in range(0, number_of_times):
    random_list = [random.randint(0, 100) for x in range (0, 50)]
    list_by_three = [a for a in random_list if a % 3 == 0]

    # start of difference calculation
    length_random = len(random_list)
    length_three = len(list_by_three)
    difference = length_random - length_three
    difference_list.append(difference)

difference_list

[34, 30, 31, 30, 31, 30, 27, 28, 35, 34]

Now that the differences are calculated, average will be calculated. 

In [14]:
average_diff = sum(difference_list)/ float(len(difference_list))

average_diff

31.0

## Activity 2: 1.02

This activity will use the first chapter of Pride and Prejudice. 

In [15]:
text = """Chapter I.
IT is a truth universally acknowledged, that a single man in possession of a good fortune must be in want of a wife.

However little known the feelings or views of such a man may be on his first entering a neighbourhood, this truth is so well fixed in the minds of the surrounding families, that he is considered as the rightful property of some one or other of their daughters.

“My dear Mr. Bennet,” said his lady to him one day, “have you heard that Netherfield Park is let at last?{2}”

Mr. Bennet replied that he had not.

“But it is,” returned she; “for Mrs. Long has just been here, and she told me all about it.”

Mr. Bennet made no answer.

“Do not you want to know who has taken it?” cried his wife, impatiently.

“You want to tell me, and I have no objection to hearing it.”


“He came down to see the place”

[Copyright 1894 by George Allen.]

This was invitation enough.

“Why, my dear, you must know, Mrs. Long says that Netherfield is taken by a young man of large fortune from the north of England; that he came down on Monday in a chaise and four to see the place, and was so much delighted with it that he agreed with Mr. Morris immediately; that he is to take possession before Michaelmas, and some of his servants are to be in the house by the end of next week.{3}”

“What is his name?”

“Bingley.”

“Is he married or single?”

“Oh, single, my dear, to be sure! A single man of large fortune; four or five thousand a year. What a fine thing for our girls!”

“How so? how can it affect them?”

“My dear Mr. Bennet,” replied his wife, “how can you be so tiresome? You must know that I am thinking of his marrying one of them.”

“Is that his design in settling here?”

“Design? Nonsense, how can you talk so! But it is very likely that he may fall in love with one of them, and therefore you must visit him as soon as he comes.”

“I see no occasion for that. You and the girls may go—or you may send them by themselves, which perhaps will be still better; for as you are as handsome as any of them, Mr. Bingley might like you the best of the party.”

“My dear, you flatter me. I certainly have had my share of beauty, but I do not pretend to be anything extraordinary now. When a woman has five grown-up daughters, she ought to give over thinking of her own beauty.”

“In such cases, a woman has not often much beauty to think of.”

“But, my dear, you must indeed go and see Mr. Bingley when he comes into the neighbourhood.”

“It is more than I engage for, I assure you.”

“But consider your daughters. Only think what an establishment it would be for one of them. Sir William and Lady Lucas are determined to go, merely on that account; for in general, you know, they visit no new{4} comers. Indeed you must go, for it will be impossible for us to visit him, if you do not.”

“You are over scrupulous, surely. I dare say Mr. Bingley will be very glad to see you; and I will send a few lines by you to assure him of my hearty consent to his marrying whichever he chooses of the girls—though I must throw in a good word for my little Lizzy.”

“I desire you will do no such thing. Lizzy is not a bit better than the others: and I am sure she is not half so handsome as Jane, nor half so good-humoured as Lydia. But you are always giving her the preference.”

“They have none of them much to recommend them,” replied he: “they are all silly and ignorant like other girls; but Lizzy has something more of quickness than her sisters.”

“Mr. Bennet, how can you abuse your own children in such a way? You take delight in vexing me. You have no compassion on my poor nerves.”

“You mistake me, my dear. I have a high respect for your nerves. They are my old friends. I have heard you mention them with consideration these twenty years at least.”

“Ah, you do not know what I suffer.”

“But I hope you will get over it, and live to see many young men of four thousand a year come into the neighbourhood.”

“It will be no use to us, if twenty such should come, since you will not visit them.”

“Depend upon it, my dear, that when there are twenty, I will visit them all.”

Mr. Bennet was so odd a mixture of quick parts, sarcastic humour, reserve, and caprice, that the experience of three-and-twenty years had been insufficient to{5} make his wife understand his character. Her mind was less difficult to develope. She was a woman of mean understanding, little information, and uncertain temper. When she was discontented, she fancied herself nervous. The business of her life was to get her daughters married: its solace was visiting and news."""

Using the text, the data type and length of the text can be found. 

In [16]:
type(text)

str

In [17]:
len(text)

4565

One objective for this activity is to count the number of times words appear. Before that can be done,  lines, unwanted spaces, special characters, and punctuation need to be removed from the text

In [18]:
# removal of lines and spacing
text = text.replace('\n', "")

In [19]:
# removal of punctuation
cleaned = ""
for char in text:
    if char == " ":
        cleaned += char
    elif char.isalnum():
        cleaned += char
    else:
        cleaned += " "

In [20]:
cleaned

'Chapter I IT is a truth universally acknowledged  that a single man in possession of a good fortune must be in want of a wife However little known the feelings or views of such a man may be on his first entering a neighbourhood  this truth is so well fixed in the minds of the surrounding families  that he is considered as the rightful property of some one or other of their daughters  My dear Mr  Bennet   said his lady to him one day   have you heard that Netherfield Park is let at last  2  Mr  Bennet replied that he had not  But it is   returned she   for Mrs  Long has just been here  and she told me all about it  Mr  Bennet made no answer  Do not you want to know who has taken it   cried his wife  impatiently  You want to tell me  and I have no objection to hearing it   He came down to see the place  Copyright 1894 by George Allen  This was invitation enough  Why  my dear  you must know  Mrs  Long says that Netherfield is taken by a young man of large fortune from the north of Englan

With the cleaned text, a list of unique words can be made, with the count. 

In [21]:
list_of_words = cleaned.split()
list_of_words

['Chapter',
 'I',
 'IT',
 'is',
 'a',
 'truth',
 'universally',
 'acknowledged',
 'that',
 'a',
 'single',
 'man',
 'in',
 'possession',
 'of',
 'a',
 'good',
 'fortune',
 'must',
 'be',
 'in',
 'want',
 'of',
 'a',
 'wife',
 'However',
 'little',
 'known',
 'the',
 'feelings',
 'or',
 'views',
 'of',
 'such',
 'a',
 'man',
 'may',
 'be',
 'on',
 'his',
 'first',
 'entering',
 'a',
 'neighbourhood',
 'this',
 'truth',
 'is',
 'so',
 'well',
 'fixed',
 'in',
 'the',
 'minds',
 'of',
 'the',
 'surrounding',
 'families',
 'that',
 'he',
 'is',
 'considered',
 'as',
 'the',
 'rightful',
 'property',
 'of',
 'some',
 'one',
 'or',
 'other',
 'of',
 'their',
 'daughters',
 'My',
 'dear',
 'Mr',
 'Bennet',
 'said',
 'his',
 'lady',
 'to',
 'him',
 'one',
 'day',
 'have',
 'you',
 'heard',
 'that',
 'Netherfield',
 'Park',
 'is',
 'let',
 'at',
 'last',
 '2',
 'Mr',
 'Bennet',
 'replied',
 'that',
 'he',
 'had',
 'not',
 'But',
 'it',
 'is',
 'returned',
 'she',
 'for',
 'Mrs',
 'Long',
 'has'

In [30]:
len(list_of_words)

871

There is a total of 871 unique words in the first chapter. 

In [22]:
unique_words = dict.fromkeys(list_of_words)
len(list(unique_words.keys()))

352

Each word will be counted to see how many times it shows up in the chapter. This will be done using a for loop. If the word shows up more than onece it will be counted/added. 

In [23]:
for word in list_of_words:
    if unique_words[word] is None:
        unique_words[word] = 1
    else: 
        unique_words[word] += 1
unique_words

{'Chapter': 1,
 'I': 18,
 'IT': 1,
 'is': 12,
 'a': 20,
 'truth': 2,
 'universally': 1,
 'acknowledged': 1,
 'that': 15,
 'single': 4,
 'man': 4,
 'in': 11,
 'possession': 2,
 'of': 29,
 'good': 3,
 'fortune': 3,
 'must': 7,
 'be': 11,
 'want': 3,
 'wife': 4,
 'However': 1,
 'little': 3,
 'known': 1,
 'the': 18,
 'feelings': 1,
 'or': 5,
 'views': 1,
 'such': 5,
 'may': 4,
 'on': 4,
 'his': 11,
 'first': 1,
 'entering': 1,
 'neighbourhood': 3,
 'this': 1,
 'so': 8,
 'well': 1,
 'fixed': 1,
 'minds': 1,
 'surrounding': 1,
 'families': 1,
 'he': 11,
 'considered': 1,
 'as': 8,
 'rightful': 1,
 'property': 1,
 'some': 2,
 'one': 5,
 'other': 2,
 'their': 1,
 'daughters': 4,
 'My': 3,
 'dear': 8,
 'Mr': 10,
 'Bennet': 6,
 'said': 1,
 'lady': 1,
 'to': 23,
 'him': 4,
 'day': 1,
 'have': 7,
 'you': 24,
 'heard': 2,
 'Netherfield': 2,
 'Park': 1,
 'let': 1,
 'at': 2,
 'last': 1,
 '2': 1,
 'replied': 3,
 'had': 3,
 'not': 9,
 'But': 6,
 'it': 11,
 'returned': 1,
 'she': 6,
 'for': 11,
 'Mrs': 

In this chapter there are many words that only show up once. With that in mind, the top 25 words will be found. These are the words that show up the most. 

In [24]:
top = sorted(unique_words.items(), key=lambda key_val_tuple: key_val_tuple[1], reverse=True)
top[:25]

[('of', 29),
 ('you', 24),
 ('to', 23),
 ('a', 20),
 ('I', 18),
 ('the', 18),
 ('and', 17),
 ('that', 15),
 ('is', 12),
 ('in', 11),
 ('be', 11),
 ('his', 11),
 ('he', 11),
 ('it', 11),
 ('for', 11),
 ('them', 11),
 ('Mr', 10),
 ('my', 10),
 ('not', 9),
 ('will', 9),
 ('so', 8),
 ('as', 8),
 ('dear', 8),
 ('was', 8),
 ('are', 8)]

The list is sorted by words that show up the most to the ones that show up less. In the first chapter, "of" shows up a total of 29 times, "are" shows up 8 times. 


## Activity 3: 2.01

permutation, iterator, lambda, list

In [28]:
permutations(range(3))

In [29]:
for number_tuple in permutations(range(3)):
    print(number_tuple)
    assert isinstance(number_tuple, tuple)

(0, 1, 2)
(0, 2, 1)
(1, 0, 2)
(1, 2, 0)
(2, 0, 1)
(2, 1, 0)


In [30]:
for number_tuple in permutations(range(3)):
    print(list(dropwhile(lambda x: x <= 0, number_tuple)))

[1, 2]
[2, 1]
[1, 0, 2]
[1, 2, 0]
[2, 0, 1]
[2, 1, 0]


In [33]:
def convert_to_number(number_stack):
    final_number = 0
    for i in range(0, len(number_stack)):
        final_number += (number_stack.pop() * (math.pow(10, i)))
    return final_number

for number_tuple in permutations(range(3)):
    number_stack = list(dropwhile(lambda x: x <= 0, number_tuple))
    print(convert_to_number(number_stack))

12.0
21.0
102.0
120.0
201.0
210.0
